apps/communications/contacts/js/bootstrap.js

```diff
@@ -0,0 +1,267 @@
+/* global applyCache */
+/* global LazyLoader */
+/* global utils */
+
+/* exports applyCache */
+
+'use strict';
+
+(function(exports) {
+
+  const FIRST_CHUNK = 'firstChunk';
+
+  // So far we only cache the first chunk of contacts, but we may want to
+  // extend the cache to a bigger or even the whole list in the future.
+  var _caches = new Map();
+  _caches.set(FIRST_CHUNK, {
+    containerId: 'groups-list',
+    active: false
+  });
+
+  var _cachedContacts;
+  var _cachedHeaders;
+
+  function setCache(aCache) {
+    if (!aCache.id || !aCache.content || !_caches.has(aCache.id)) {
+      return;
+    }
+
+    localStorage.setItem(aCache.id, JSON.stringify(aCache.content));
+  }
+
+  function deleteCache(aId) {
+    localStorage.removeItem(aId);
+  }
+
+  function getCachedContacts(aNodeList) {
+    if (!aNodeList) {
+      return;
+    }
+
+    if (!_cachedContacts) {
+      _cachedContacts = new Map();
+    }
+
+    for (var i = 0; i < aNodeList.length; i++) {
+      _cachedContacts.set(aNodeList[i].dataset.uuid,
+                          aNodeList[i].innerHTML);
+    }
+  }
+
+  function appendNodesToContainer(aContainer, aNodeList) {
+    if (!aNodeList.length) {
+      return Promise.resolve();
+    }
+
+    return new Promise((resolve) => {
+      var resolved = false;
+      var fragment = document.createDocumentFragment();
+      aNodeList.forEach((node) => {
+        if (!node.elementName) {
+          return;
+        }
+        var child = document.createElement(node.elementName);
+        node.attributes && node.attributes.forEach((attribute) => {
+          if (!attribute.name || !attribute.value) {
+            return;
+          }
+          child.setAttribute(attribute.name, attribute.value);
+        });
+        if (node.innerHTML) {
+          child.innerHTML = node.innerHTML;
+        }
+        fragment.appendChild(child);
+
+        var headerName = child.id.split('section-group-')[1];
+        if (!_cachedHeaders) {
```

Comment:
nit: we should check for the existence of `_cachedHeaders` outside of the forEach loop